cpp/include/raft/detail/span.hpp

```diff
@@ -86,5 +86,30 @@ __host__ __device__ constexpr auto lexicographical_compare(InputIt1 first1,
   }
   return first1 == last1 && first2 != last2;
 }
+
+template <typename T, typename SizeType, size_t Extent>
+struct span_storage {
+ private:
+  T* ptr_{nullptr};
+
+ public:
+  constexpr span_storage() = default;
+  constexpr span_storage(T* ptr, SizeType) : ptr_{ptr} {}
+  [[nodiscard]] constexpr auto size() const -> SizeType { return Extent; }
+  constexpr auto data() const -> T* { return ptr_; }
+};
+
+template <typename T, typename SizeType>
+struct span_storage<T, SizeType, dynamic_extent> {
+ private:
+  T* ptr_{nullptr};
+  SizeType size_{0};
+
+ public:
+  constexpr span_storage() = default;
+  constexpr span_storage(T* ptr, SizeType size) : ptr_{ptr}, size_{size} {}
+  [[nodiscard]] constexpr auto size() const -> SizeType { return size_; }
+  constexpr auto data() const -> T* { return ptr_; }
```

Comment:
```suggestion
  constexpr span_storage() noexcept = default;
  constexpr span_storage(T* ptr, SizeType size) noexcept : ptr_{ptr}, size_{size} {}
  [[nodiscard]] constexpr auto size() const noexcept -> SizeType { return size_; }
  constexpr auto data() const noexcept -> T* { return ptr_; }
```

Reply:
Added `noexcept` for all functions.  Thank you for the suggestion.